alf/environments/alf_gym3_wrapper.py

```diff
@@ -0,0 +1,407 @@
+            prev_action=torch.as_tensor(action),
+            env_info=nest.map_structure(
+                lambda *values: torch.as_tensor(values), *trimmed_info))
+
+    # Implement abstract _reset
+    def _reset(self) -> ds.TimeStep:
+        """Implement ``_reset()`` for compatibility
+
+        Note that by default Gym3 environmenst cannot be reset. However, if
+        ``support_force_reset`` is set to True, action -1 will be sent to all
+        the sub-environments to reset them if the underlying environment follows
+        the convention.
+
+        Otherwise the reset will be ignored.
+
+        """
+        if self._support_force_reset:
+            self._gym3_env.act(np.array([-1] * self.batch_size))
+            self._done = [False] * self.batch_size
+        else:
+            logging.warning('reset() ignored by AlfGym3Wrapper')
+
+        reward, observation, _ = self._gym3_env.observe()
+
+        time_step = self._create_time_step(
+            reward,
+            observation,
+            step_type=[ds.StepType.FIRST] * self.batch_size,
+            # Faking actions
+            action=nest.map_structure(
+                lambda spec: spec.numpy_zeros(outer_dims=(self.batch_size, )),
+                self.action_spec()))
+
+        return time_step
+
+    # Implement abstract _step
+    def _step(self, action) -> ds.TimeStep:
+        """Implement ``_step()`` for compatibility
+
+        A very important note here is that special treatment is done at the end
+        of each episode. Unlike Gym environments, Gym3 environments do not
+        return ``done=True`` immediately. Instead, ``first=True`` is returned
+        upon the ``observe()`` of the NEXT ``act()``, which is actually the
+        FIRST FRAME of a new episode in Gym3 sense.
+
+        With this wrapper we CHANGED that definition, so that the frame with
+        ``first=True`` actually becomes the LAST FRAME of the previous episode.
+        Because the observation of the new episode can be DRAMATICALLY DIFFERENT
+        from the actual last frame of the previous episode, the previous
+        observation is returned in this case.
+
+        Therefore to summarize, effectively we will
+
+        1. Repeat the end-of-episode observation twice for each episode.
+
+        2. Throw away the first frame of each episode, and use the second frame
+           as if it is the first frame.
+
+        Gym3's official Gym wrapper DID THE SAME.
+
+        """
+        _, prev_observation, _ = self._gym3_env.observe()
+
+        np_action = nest.map_structure(lambda x: x.cpu().numpy(), action)
+        self._gym3_env.act(np_action)
+
+        reward, observation, first = self._gym3_env.observe()
+
+        # Override the obervation with the previous observation if that
+        # particular environment has ``first=True``.
+        for i in range(self.batch_size):
+            if first[i]:
+                observation[i] = prev_observation[i]
```

Comment:
I think the type of observation  can be  `Dict[str, np.ndarray]`. The unittest should test both Tensor observation space and dict observation space.

Reply:
How about this one?

Reply:
Yes the observation could be of type dictionary. But instead of the normal `Dict`, it was `gym3`'s `DictType`, which is not recognized by `nest` as nested structure.

We can easily convert `DictType` to `Dict` if it only appears at the top level of observation structure, but I haven't found that assumption being mentioned yet.

Reply:
If DictType cannot be handled,  map_structure at line 285 and 362 will not do the right job.

Reply:
Yep I realized that. I am thinking about the best way to handle it. The quote above seem to suggest that only the first level in observation/action could be `DictType`. Can we note this assumption and handle accordingly?

Reply:
Perhaps can do something similar to the _gym_space_to_nested_space() in gym_wrappers.py

Reply:
This is now handled assuming the first layer of would be a `DictType` in `gym3`. Also updated the unit test to remove the `ExtractDictObWrapper`. This makes sure that both cases are tested because:

1. `observation_spec` in this case is a `DictType[Discrete Tensor]`, while
2. `action_spec` is a `Discrete Tensor`.

Reply:
I am wondering if observation is a dict, what is meaning of observation[i]?

Reply:
Getting sloppy here because the unit test isn't exactly covering this case. In my original unit test it did not properly test the intermediate `FIRST` frame case. Let me find a way to actually test this.

Reply:
This is now fixed and added unit test to explicitly test such special handling. PTAL. Thanks for catching this!